File: Windows/BibleVerseReplacer.Windows/ArticleReferenceReplacer.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace BibleVerseReplacer.Windows
{
    internal sealed class ArticleReplacementResult
    {
        public ArticleReplacementResult(string text, int replacements, int skippedExisting)
        {
            Text = text;
            Replacements = replacements;
            SkippedExisting = skippedExisting;
        }

        public string Text { get; private set; }
        public int Replacements { get; private set; }
        public int SkippedExisting { get; private set; }

        public bool Changed
        {
            get { return Replacements > 0; }
        }
    }

    internal sealed class ArticleReferenceReplacer
    {
        private readonly ReferenceParser parser;
        private readonly VerseFormatter formatter;

        public ArticleReferenceReplacer(ReferenceParser parser, VerseFormatter formatter)
        {
            this.parser = parser;
            this.formatter = formatter;
        }

        public ArticleReplacementResult ReplaceReferences(
            string article,
            OutputFormat format,
            ReferenceLabelMode labelMode,
            CombinedPassageMode combinedPassageMode,
            QuotationStyle quotationStyle)
        {
            StringBuilder builder = new StringBuilder();
            int cursor = 0;
            int replacements = 0;
            int skipped = 0;

            while (cursor < article.Length)
            {
                Candidate candidate;
                if (!NextCandidate(article, cursor, out candidate))
                {
                    builder.Append(article.Substring(cursor));
                    break;
                }

                builder.Append(article.Substring(cursor, candidate.Start - cursor));

                try
                {
                    ParsedReference reference = parser.ParseSelection(candidate.Raw);
                    List<PassageVerseGroup> verseGroups = BibleStore.Instance.VerseGroupsFor(reference);
                    List<BibleVerse> verses = BibleStore.Instance.VersesFor(reference);
                    if (ScriptureAlreadyPresent(article, candidate.End, verses, quotationStyle))
                    {
                        builder.Append(article.Substring(candidate.Start, candidate.End - candidate.Start));
                        skipped++;
                    }
                    else
                    {
                        builder.Append(formatter.Format(
                            reference,
                            verses,
                            verseGroups,
                            format,
                            labelMode,
                            candidate.Raw,
                            combinedPassageMode,
                            quotationStyle));
                        replacements++;
                    }
                    cursor = candidate.End;
                }
                catch
                {
                    builder.Append(article[candidate.Start]);
                    cursor = candidate.Start + 1;
                }
            }

            return new ArticleReplacementResult(builder.ToString(), replacements, skipped);
        }

        private static bool NextCandidate(string text, int start, out Candidate candidate)
        {
            for (int index = start; index < text.Length; index++)
            {
                if (IsReferenceStart(text, index))
                {
                    int end = CandidateEnd(text, index);
                    int trimmedEnd = TrimCandidateEnd(text, index, end);
                    if (trimmedEnd > index)
                    {
                        candidate = new Candidate(index, trimmedEnd, text.Substring(index, trimmedEnd - index));
                        return true;
                    }
                }
            }

            candidate = null;
            return false;
        }

        private static bool IsReferenceStart(string text, int index)
        {
            if (index > 0 && (char.IsLetterOrDigit(text[index - 1])))
            {
                return false;
            }

            BibleBook book;
            string remaining;
            if (!BibleBookCatalog.FindAtStart(text.Substring(index), out book, out remaining))
            {
                return false;
            }

            int lookahead = index + RawPrefixLength(book, text.Substring(index));
            while (lookahead < text.Length && char.IsWhiteSpace(text[lookahead]))
            {
                lookahead++;
            }
            return lookahead < text.Length && (char.IsDigit(text[lookahead]) || text[lookahead] == '第');
        }

        private static int RawPrefixLength(BibleBook book, string text)
        {
            List<string> aliases = new List<string>();
            aliases.Add(book.ChineseName);
            aliases.Add(book.Code);
            aliases.AddRange(book.Aliases);
            aliases.Sort((left, right) => right.Length.CompareTo(left.Length));
            foreach (string alias in aliases)
            {
                if (text.StartsWith(alias, StringComparison.OrdinalIgnoreCase))
                {
                    return alias.Length;
                }
            }
            return book.ChineseName.Length;
        }

        private static int CandidateEnd(string text, int start)
        {
            BibleBook book;
            string remaining;
            if (!BibleBookCatalog.FindAtStart(text.Substring(start), out book, out remaining))
            {
                return start;
            }

            int index = start + RawPrefixLength(book, text.Substring(start));
            while (index < text.Length && IsCandidateCharacter(text[index]))
            {
                index++;
            }
            return index;
        }

        private static bool IsCandidateCharacter(char ch)
        {
            if (char.IsDigit(ch) || char.IsWhiteSpace(ch))
            {
                return true;
            }
            if ("toTO".IndexOf(ch) >= 0)
            {
                return true;
            }
            return ":：﹕,，、;；|｜\\-－–—―﹣～~^.…第章节到至".IndexOf(ch) >= 0;
        }

        private static int TrimCandidateEnd(string text, int start, int end)
        {
            int trimmed = end;
            while (trimmed > start)
            {
                char previous = text[trimmed - 1];
                if (char.IsWhiteSpace(previous) || ",，、;；|｜\\".IndexOf(previous) >= 0)
                {
                    trimmed--;
                }
                else
                {
                    break;
                }
            }
            return trimmed;
        }

        private static bool ScriptureAlreadyPresent(
            string article,
            int index,
            IList<BibleVerse> verses,
            QuotationStyle quotationStyle)
        {
            if (verses.Count == 0)
            {
                return false;
            }

            int cursor = index;
            while (cursor < article.Length && (char.IsWhiteSpace(article[cursor]) || ":：".IndexOf(article[cursor]) >= 0))
            {
                cursor++;
            }

            string remaining = article.Substring(cursor);
            string expected = VerseFormatter.CleanText(verses[0].Text, quotationStyle);
            return NormalizedScripturePrefix(remaining).StartsWith(
                NormalizedScripturePrefix(expected.Length > 12 ? expected.Substring(0, 12) : expected),
                StringComparison.Ordinal);
        }

        private static string NormalizedScripturePrefix(string text)
        {
            StringBuilder builder = new StringBuilder();
            int limit = Math.Min(30, text.Length);
            for (int index = 0; index < limit; index++)
            {
                char ch = text[index];
                if (char.IsWhiteSpace(ch))
                {
                    continue;
                }
                if (ch == '「' || ch == '」' || ch == '“' || ch == '”')
                {
                    builder.Append('"');
                }
                else
                {
                    builder.Append(ch);
                }
            }
            return builder.ToString();
        }

        private sealed class Candidate
        {
            public Candidate(int start, int end, string raw)
            {
                Start = start;
                End = end;
                Raw = raw;
            }

            public int Start { get; private set; }
            public int End { get; private set; }
            public string Raw { get; private set; }
        }
    }
}

File: Sources/BibleVerseReplacer/ArticleReferenceReplacer.swift
import Foundation

struct ArticleReplacementResult {
    let text: String
    let replacements: Int
    let skippedExisting: Int

    var changed: Bool {
        replacements > 0
    }
}

final class ArticleReferenceReplacer {
    private let parser: ReferenceParser
    private let formatter: VerseFormatter
    private let bibleStore: BibleStore

    init(
        parser: ReferenceParser = ReferenceParser(),
        formatter: VerseFormatter = VerseFormatter(),
        bibleStore: BibleStore = .shared
    ) {
        self.parser = parser
        self.formatter = formatter
        self.bibleStore = bibleStore
    }

    func replaceReferences(
        in article: String,
        format: OutputFormat,
        labelMode: ReferenceLabelMode,
        combinedPassageMode: CombinedPassageMode,
        quotationStyle: QuotationStyle
    ) -> ArticleReplacementResult {
        var result = ""
        var cursor = article.startIndex
        var replacements = 0
        var skipped = 0

        while cursor < article.endIndex {
            guard let candidate = nextCandidate(in: article, from: cursor) else {
                result.append(contentsOf: article[cursor...])
                break
            }

            result.append(contentsOf: article[cursor..<candidate.range.lowerBound])

            do {
                let reference = try parser.parseSelection(candidate.raw)
                let verseGroups = try bibleStore.verseGroups(for: reference)
                let verses = try bibleStore.verses(for: reference)
                if scriptureAlreadyPresent(after: candidate.range.upperBound, in: article, verses: verses, quotationStyle: quotationStyle) {
                    result.append(contentsOf: article[candidate.range])
                    skipped += 1
                } else {
                    result.append(formatter.format(
                        parsedReference: reference,
                        verses: verses,
                        verseGroups: verseGroups,
                        format: format,
                        labelMode: labelMode,
                        originalReference: candidate.raw,
                        combinedPassageMode: combinedPassageMode,
                        quotationStyle: quotationStyle
                    ))
                    replacements += 1
                }
                cursor = candidate.range.upperBound
            } catch {
                result.append(article[candidate.range.lowerBound])
                cursor = article.index(after: candidate.range.lowerBound)
            }
        }

        return ArticleReplacementResult(text: result, replacements: replacements, skippedExisting: skipped)
    }

    private func nextCandidate(in text: String, from start: String.Index) -> (range: Range<String.Index>, raw: String)? {
        var index = start
        while index < text.endIndex {
            if isReferenceStart(in: text, at: index),
               let end = candidateEnd(in: text, from: index) {
                let trimmedEnd = trimCandidateEnd(in: text, start: index, end: end)
                if trimmedEnd > index {
                    return (index..<trimmedEnd, String(text[index..<trimmedEnd]))
                }
            }
            index = text.index(after: index)
        }
        return nil
    }

    private func isReferenceStart(in text: String, at index: String.Index) -> Bool {
        if text[index].isWhitespace {
            return false
        }

        if index > text.startIndex {
            let previous = text[text.index(before: index)]
            if previous.isNumber || previous.isASCIILetter {
                return false
            }
        }

        guard let match = BibleBookCatalog.bookAtStart(of: String(text[index...])) else {
            return false
        }

        var lookahead = text.index(index, offsetBy: rawPrefixLength(for: match.book, in: text[index...]), limitedBy: text.endIndex) ?? index
        while lookahead < text.endIndex, text[lookahead].isWhitespace {
            lookahead = text.index(after: lookahead)
        }
        guard lookahead < text.endIndex else {
            return false
        }
        return text[lookahead].isNumber || text[lookahead] == "第"
    }

    private func rawPrefixLength(for book: BibleBook, in text: Substring) -> Int {
        let aliases = ([book.chineseName, book.code] + book.aliases).sorted { $0.count > $1.count }
        let lower = text.lowercased()
        for alias in aliases where lower.hasPrefix(alias.lowercased()) {
            return alias.count
        }
        return book.chineseName.count
    }

    private func candidateEnd(in text: String, from start: String.Index) -> String.Index? {
        guard let match = BibleBookCatalog.bookAtStart(of: String(text[start...])) else {
            return nil
        }

        var index = text.index(
            start,
            offsetBy: rawPrefixLength(for: match.book, in: text[start...]),
            limitedBy: text.endIndex
        ) ?? start
        while index < text.endIndex {
            let character = text[index]
            if !isCandidateCharacter(character) {
                break
            }
            index = text.index(after: index)
        }
        return index > start ? index : nil
    }

    private func isCandidateCharacter(_ character: Character) -> Bool {
        if character.isNumber || character.isWhitespace {
            return true
        }
        if "toTO".contains(character) {
            return true
        }
        return ":：﹕,，、;；|｜\\-－–—―﹣～~^.…第章节到至".contains(character)
    }

    private func trimCandidateEnd(in text: String, start: String.Index, end: String.Index) -> String.Index {
        var trimmed = end
        while trimmed > start {
            let previous = text[text.index(before: trimmed)]
            if previous.isWhitespace || ",，、;；|｜\\".contains(previous) {
                trimmed = text.index(before: trimmed)
            } else {
                break
            }
        }
        return trimmed
    }

    private func scriptureAlreadyPresent(
        after index: String.Index,
        in text: String,
        verses: [BibleVerse],
        quotationStyle: QuotationStyle
    ) -> Bool {
        guard let firstVerse = verses.first else {
            return false
        }

        var cursor = index
        while cursor < text.endIndex, text[cursor].isWhitespace || ":：".contains(text[cursor]) {
            cursor = text.index(after: cursor)
        }

        let remaining = String(text[cursor...])
        let expected = formatter.displayText(for: firstVerse, quotationStyle: quotationStyle)
        return normalizedScripturePrefix(remaining).hasPrefix(normalizedScripturePrefix(String(expected.prefix(12))))
    }

    private func normalizedScripturePrefix(_ text: String) -> String {
        text
            .prefix(30)
            .map { character -> Character in
                switch character {
                case "「", "“":
                    return "\""
                case "」", "”":
                    return "\""
                default:
                    return character
                }
            }
            .filter { !$0.isWhitespace }
            .map(String.init)
            .joined()
    }
}

private extension Character {
    var isASCIILetter: Bool {
        unicodeScalars.count == 1 && unicodeScalars.allSatisfy { scalar in
            (65...90).contains(scalar.value) || (97...122).contains(scalar.value)
        }
    }
}
